### PR TITLE
[SYCL][Doc] Add ESIMD design and implementation notes

### PR DESCRIPTION
**Pull request by kbobrovs**
Signed-off-by: Konstantin S Bobrovsky <konstantin.s.bobrovsky@intel.com>

### CODE REVIEW

File: sycl/doc/design/ESIMDDesignNotes.md [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+
+## Directions
+
+This section lists possible directions for ESIMD improvements.
+
+1. Support `std::simd`. This is the standard C++ way for explicit SIMD programming. Can help run (subsest of ESIMD) on CPU efficiently in the future.
+1. Clear (via namespace?) separation of ESIMD APIs into portable and
+architecture-specific parts.
+1. Standardizing `simd_view` or equivalent. This is effectively a reference to a *subset* of `esimd::simd` vector object�s elements. The subset is defined in a regular way via starting offset, stride and number of elements in the subset. This proved to be very useful and loved by users. Missing in `std::simd`.
```
Comment (kbobrovs) [inline sycl/doc/design/ESIMDDesignNotes.md:231]:
```suggestion
1. Standardizing `simd_view` or equivalent. This is effectively a reference
    to a *subset* of `esimd::simd` vector object's elements. The subset is defined
    in a regular way via starting offset, stride and number of elements in the subset.
    This proved to be very useful and loved by users. Missing in `std::simd`.
```

File: sycl/doc/design/ESIMDDesignNotes.md [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+To facilitate this splitting, compiler will automatically identify markup and
+clone parts of the ESIMD callgraph starting from kernels and functions
+explicitly marked with the `intel::sycl_explicit_simd` attribute.
+
+### Device headers
+There are two distinct parts of ESIMD support in the SYCL device headers. The
+first one is ESIMD-related "tweaks" within the usual SYCL headers, and the
+second is the ESIMD APIs themselves.
+
+##### ESIMD tweaks in SYCL device headers
```
Comment (kbobrovs) [inline sycl/doc/design/ESIMDDesignNotes.md:25]:
```suggestion
#### ESIMD tweaks in SYCL device headers
```

File: sycl/doc/design/ESIMDDesignNotes.md [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+There are also couple ESIMD-specific proxy classes used by ESIMD API 
+implementations to access internals of ESIMD objects:
+- `sycl::ext::intel::esimd::detail::AccessorPrivateProxy`
+this one is used by ESIMD memory APIs to access internals of `sycl::accessor`
+objects
+- `sycl::ext::intel::esimd::detail::WrapperElementTypeProxy`
+used to access internals of SYCL types such as
+`sycl::detail::half_impl::half`
+
+##### ESIMD API headers
```
Comment (kbobrovs) [inline sycl/doc/design/ESIMDDesignNotes.md:44]:
```suggestion
#### ESIMD API headers
```

File: sycl/doc/design/ESIMDDesignNotes.md [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+<tt>clang/lib/Sema/SemaExpr.cpp</tt><br>
+<tt>clang/include/clang/Basic/Attr.td</tt><br>
+<tt>clang/include/clang/Basic/DiagnosticSemaKinds.td</tt><br>
+
+#### Code (LLVMIR) generator
+
+ESIMD-specific code generator tweaks are mostly translations of internal FE 
+representation of variaous ESIMD attributes into LLVM IR attributes or metadata.
+
+###### Kernel signature generation
```
Comment (kbobrovs) [inline sycl/doc/design/ESIMDDesignNotes.md:99]:
```suggestion
##### Kernel signature generation
```

File: sycl/doc/design/ESIMDDesignNotes.md [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+For ESIMD kernes, a number of additional attributes are generated for the kernel
+function itself as well as certain argument.
+- Kernels are annotated with `sycl_explicit_simd` and
+`intel_reqd_sub_group_size` attributes. The latter must always be `1` for
+a ESIMD kernel or function.
+- An argument which conveys accessor's pointer is assigned a `kernel_arg_accessor_ptr` attribute
+
+###### Global variable code generation.
```
Comment (kbobrovs) [inline sycl/doc/design/ESIMDDesignNotes.md:107]:
```suggestion
##### Global variable code generation.
```

File: sycl/doc/design/ESIMDDesignNotes.md [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+ESIMD supports "private globals" - global variables which have one copy per
+thread of execution (similar to C++ thread_local), normally allocated of Gen
+register file. To make a global variable a "private global", ` __attribute__((opencl_private)) __attribute__((sycl_explicit_simd))` attributes are
+used. Globals of this can be forced to a specific register using the
+`__attribute__((register_num(n)))` attribute. The clang code generator
+translates these to `genx_volatile` and `genx_byte_offset` LLVM IR
+attributes.
+
+###### Function attributes translations:
```
Comment (kbobrovs) [inline sycl/doc/design/ESIMDDesignNotes.md:116]:
```suggestion
##### Function attributes translations:
```

**Review by v-klochkov — APPROVED** [review]

File: sycl/doc/design/ESIMDDesignNotes.md [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+  inclusion of the integration header generated by the generator. For ESIMD
+  kernels `isESIMD()` returns true, and this is used in SYCL RT headers to
+  distinguish ESIMD kernels from SYCL kernels, e.g. to customize kernel argument
+  setting for accessors.
+- defines `sycl_esimd_vectorize` attribute propagated to the VC BE, aimed at
+  making it possible to call SIMT function from ESIMD code while vectorizing the
+  former. Used internally to implement certain math functions in SYCL spec
+  conformant manner (proper precision) via vectorizing scalar conformant
+  implementations - see example
+  [usage](https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/intel/esimd/detail/math_intrin.hpp#L199)
```
Comment (bader) [inline sycl/doc/design/ESIMDDesignNotes.md:97]: I think we should move this type of documentation the source code and use Doxygen to produce rendered version. It's much more convenient to track the correlation between the code and the documentation. The links and source location pointers are easy to invalidate with code changes.

File: sycl/doc/design/ESIMDDesignNotes.md [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+post-link tool splits the input module (or modules resulting from splitting
+by other characteristics, such as aspects) into two - SYCL and ESIMD ones.
+Shared functions invoked both from SYCL and ESIMD are cloned during the process.
+This is necessary because SYCL and ESIMD parts must undergo different set of
+transformations before generating resulting SPIR-V. ESIMD modules resulting from
+splitting are marked with specific device binary property `isEsimdImage` (see
+[source](https://github.com/intel/llvm/blob/9dc14a292f479880b5ab2e95f99a8414b31d1165/llvm/tools/sycl-post-link/sycl-post-link.cpp#L453)
+.)
+
+#### Driver
```
Comment (bader) [inline sycl/doc/design/ESIMDDesignNotes.md:169]: This section should be renamed. We already have another component with the name driver and any other document references `sycl-post-link` as a driver. Probably we can merge this sub-section the parent section above.
@v-klochkov, what do you think?

**Review by bader — APPROVED** [review]

File: sycl/doc/design/ESIMDDesignNotes.md (hunk 8 above)
Comment (bader) [reply sycl/doc/design/ESIMDDesignNotes.md]: Done in https://github.com/intel/llvm/pull/7824/commits/1b3a19f059878edb03ce1314b28dbd4d9dfc1871. Please, let me know if it looks good to you.

Comment (v-klochkov) [reply sycl/doc/design/ESIMDDesignNotes.md]: Looks good to me. Thank you for doing these additional changes!

**Pull request merged**